File: macos/Tests/WordingKitTests/PackSlugTests.swift
import Testing

@testable import WordingKit

/// The identifier of a downloaded pack decides which file gets written, so these are
/// the checks standing between an arbitrary URL and the user's data directory.
@Suite struct PackSlugTests {
    @Test(arguments: [
        ("travel-basics", "travel-basics"),
        ("Travel-Basics", "travel-basics"),
        ("a", "a"),
        ("es2000", "es2000"),
    ])
    func acceptsPlainIdentifiers(id: String, expected: String) {
        #expect(PackSlug.normalize(id) == expected)
    }

    @Test(arguments: [
        "../words", "../../words", "..", ".", "sets/../words",
        "a/b", "a\\b", "/etc/passwd", "C:\\words", "words.json",
    ])
    func refusesAnythingThatCouldChooseItsOwnPath(id: String) {
        #expect(PackSlug.normalize(id) == nil)
    }

    @Test(arguments: ["con", "CON", "com1", "nul", "lpt9"])
    func refusesNamesWindowsReserves(id: String) {
        // All letters and digits, so the character rule alone would let them through.
        // Checked on macOS too: a pack accepted here has to be storable on Windows.
        #expect(PackSlug.normalize(id) == nil)
    }

    @Test(arguments: ["", " ", "-leading", "trailing-", "with space", "zażółć", "emoji-😀"])
    func refusesEverythingOutsideTheAllowList(id: String) {
        #expect(PackSlug.normalize(id) == nil)
    }

    @Test func refusesAnIdentifierLongerThanTheLimit() {
        #expect(PackSlug.normalize(String(repeating: "a", count: PackLimits.maxIdLength)) != nil)
        #expect(PackSlug.normalize(String(repeating: "a", count: PackLimits.maxIdLength + 1)) == nil)
    }

    @Test func requireThrowsTheSharedPackError() {
        #expect(throws: WordPackError.unsafeId) { try PackSlug.require("../escape") }
    }
}

File: tests/Wording.Core.Tests/PackSlugTests.cs
using Wording.Core.Packs;

namespace Wording.Core.Tests;

/// <summary>
/// The identifier of a downloaded pack decides which file gets written, so these are
/// the checks standing between an arbitrary URL and the user's data directory.
/// </summary>
public class PackSlugTests
{
    [Theory]
    [InlineData("travel-basics", "travel-basics")]
    [InlineData("Travel-Basics", "travel-basics")]
    [InlineData("a", "a")]
    [InlineData("es2000", "es2000")]
    public void TryNormalize_AcceptsPlainIdentifiers(string id, string expected)
    {
        Assert.True(PackSlug.TryNormalize(id, out var slug));
        Assert.Equal(expected, slug);
    }

    [Theory]
    [InlineData("../words")]
    [InlineData("../../words")]
    [InlineData("..")]
    [InlineData(".")]
    [InlineData("sets/../words")]
    [InlineData("a/b")]
    [InlineData("a\\b")]
    [InlineData("/etc/passwd")]
    [InlineData("C:\\words")]
    [InlineData("words.json")]
    public void TryNormalize_RefusesAnythingThatCouldChooseItsOwnPath(string id)
    {
        Assert.False(PackSlug.TryNormalize(id, out _));
    }

    [Theory]
    [InlineData("con")]
    [InlineData("CON")]
    [InlineData("com1")]
    [InlineData("nul")]
    [InlineData("lpt9")]
    public void TryNormalize_RefusesNamesWindowsReserves(string id)
    {
        // All letters and digits, so the character rule alone would let them through.
        Assert.False(PackSlug.TryNormalize(id, out _));
    }

    [Theory]
    [InlineData("")]
    [InlineData(" ")]
    [InlineData("-leading")]
    [InlineData("trailing-")]
    [InlineData("with space")]
    [InlineData("zażółć")]
    [InlineData("emoji-\U0001F600")]
    [InlineData("null\0byte")]
    public void TryNormalize_RefusesEverythingOutsideTheAllowList(string id)
    {
        Assert.False(PackSlug.TryNormalize(id, out _));
    }

    [Fact]
    public void TryNormalize_RefusesAnIdentifierLongerThanTheLimit()
    {
        Assert.True(PackSlug.TryNormalize(new string('a', PackLimits.MaxIdLength), out _));
        Assert.False(PackSlug.TryNormalize(new string('a', PackLimits.MaxIdLength + 1), out _));
    }

    [Fact]
    public void Require_ThrowsTheSharedPackError()
    {
        var error = Assert.Throws<WordPackException>(() => PackSlug.Require("../escape"));

        Assert.Equal(PackProblem.UnsafeId, error.Problem);
    }
}
